finagle-redis/src/main/scala/com/twitter/finagle/Redis.scala

```diff
@@ -63,4 +66,48 @@ object Redis extends Client[Command, Reply] {
 
   def newService(dest: Name, label: String): Service[Command, Reply] =
     client.newService(dest, label)
+
+  object Subscribe extends com.twitter.finagle.Client[SubscribeCommand, Unit] {
+
+    object SubscribeClient {
+      /**
+       * Default stack parameters used for redis client.
+       */
+      val defaultParams: Stack.Params = StackClient.defaultParams +
+        param.ProtocolLibrary("redis.subscribe")
+
+      /**
+       * A default client stack which supports the pipelined redis client.
+       */
+      def newStack: Stack[ServiceFactory[SubscribeCommand, Unit]] = StackClient.newStack
+        .replace(DefaultPool.Role, SingletonPool.module[SubscribeCommand, Unit])
+    }
+
+    case class SubscribeClient(
```

Comment:
Some more thoughts on adding subscribe support to the existing pipelining client:

1. It is hard to tell whether a client has quit from pub/sub mode successfully, because we are not replied when we issue an unsubscribe command. When we issued  an unsubscribe command to the last channel/pattern we are currently subscribed to, we have to wait until the acknowledgement message for this command arrived, before we can confidently say that it is safe to send normal commands. But during the waiting gap, it is possible that some other subscribe command are issued. So, the acknowledgement message carries some information about the connection status at the time it is sent out from the server, which may not still be true at the time it is received by the client. To solve this problem, we'll have to provide some special unsubscribe command, so that when it is issued, the client put itself into a temporally-unavailable state, and reject any more command, neither normal nor subscribe, until it receives the unsubscribe acknowledgement.

2. We cannot manage the client pub/sub status transparently with sharding. We'll have to provider sharders specifically designed for redis, to keep track all of the pub/sub status of the underlying shards.